src/core/internal/traits.d

```diff
@@ -303,6 +305,110 @@ template hasElaborateAssign(S)
     }
 }
 
+template hasIndirections(T)
+{
+    static if (is(T == struct) || is(T == union))
+        enum hasIndirections = anySatisfy!(.hasIndirections, Fields!T);
+    else static if (__traits(isStaticArray, T) && is(T : E[N], E, size_t N))
+        enum hasIndirections = is(E == void) ? true : hasIndirections!E;
+    else static if (isFunctionPointer!T)
+        enum hasIndirections = false;
+    else
+        enum hasIndirections = isPointer!T || isDelegate!T || isDynamicArray!T ||
+            __traits(isAssociativeArray, T) || is (T == class) || is(T == interface);
+}
+
+template hasUnsharedIndirections(T)
+{
+    static if (is(T == struct) || is(T == union))
+        enum hasUnsharedIndirections = anySatisfy!(.hasUnsharedIndirections, Fields!T);
+    else static if (is(T : E[N], E, size_t N))
+        enum hasUnsharedIndirections = is(E == void) ? false : hasUnsharedIndirections!E;
+    else static if (isFunctionPointer!T)
+        enum hasUnsharedIndirections = false;
+    else static if (isPointer!T)
+        enum hasUnsharedIndirections = !is(T : shared(U)*, U);
+    else static if (isDynamicArray!T)
+        enum hasUnsharedIndirections = !is(T : shared(V)[], V);
+    else static if (is(T == class) || is(T == interface))
+        enum hasUnsharedIndirections = !is(T : shared(W), W);
+    else
+        enum hasUnsharedIndirections = isDelegate!T || __traits(isAssociativeArray, T); // TODO: how to handle these?
+}
+
+enum bool isAggregateType(T) = is(T == struct) || is(T == union) ||
+                               is(T == class) || is(T == interface);
+
+enum bool isPointer(T) = is(T == U*, U) && !isAggregateType!T;
+
+enum bool isDynamicArray(T) = is(DynamicArrayTypeOf!T) && !isAggregateType!T;
+
+template OriginalType(T)
+{
+    template Impl(T)
+    {
+        static if (is(T U == enum)) alias Impl = OriginalType!U;
+        else                        alias Impl =              T;
+    }
+
+    alias OriginalType = ModifyTypePreservingTQ!(Impl, T);
+}
+
+template DynamicArrayTypeOf(T)
+{
+    static if (is(AliasThisTypeOf!T AT) && !is(AT[] == AT))
+        alias X = DynamicArrayTypeOf!AT;
+    else
+        alias X = OriginalType!T;
+
+    static if (is(Unqual!X : E[], E) && !is(typeof({ enum n = X.length; })))
+        alias DynamicArrayTypeOf = X;
+    else
+        static assert(0, T.stringof ~ " is not a dynamic array");
+}
+
+private template AliasThisTypeOf(T)
+    if (isAggregateType!T)
```

Comment:
Shouldn't be indented.

Reply:
Really? Every constraint I've seen has been intended... that's how you know it's a constraint.

Reply:
[Yes](https://dlang.org/dstyle.html#phobos_declarations), "Constraints on declarations should have the same indentation level as their declaration".

Reply:
1. that's horrible.
2. it's already a 50/50 split in druntime. i think it actually leans in favour of this way, which is why i went with this way.

Reply:
AIUI, it's only split because the style guide was written/followed after a lot of the code had already been already written. I'm not sure I agree with this rule either, but it's required by the style guide...

Reply:
Well someone can produce a PR that fixes all the code in one hit.